### ISSUE TITLE
Support for authSource for mongodb

### ISSUE THREAD

**Issue by bwgjoseph** [hariprasanths/synchly#3]
It seem like if the database has `authSource` enabled, it will not be able to connect

```
× Authentication failed
Database configuration update failed.
MongoError: Authentication failed.
Re run with --config db to finish the configuration
Stacktrace:
MongoError: Authentication failed.
    at MessageStream.messageHandler (C:\Users\Joseph\AppData\Roaming\npm\node_modules\synchly\node_modules\mongodb\lib\cmap\connection.js:266:20)
    at MessageStream.emit (events.js:310:20)
    at processIncomingData (C:\Users\Joseph\AppData\Roaming\npm\node_modules\synchly\node_modules\mongodb\lib\cmap\message_stream.js:144:12)
    at MessageStream._write (C:\Users\Joseph\AppData\Roaming\npm\node_modules\synchly\node_modules\mongodb\lib\cmap\message_stream.js:42:5)
    at doWrite (_stream_writable.js:403:12)
    at writeOrBuffer (_stream_writable.js:387:5)
    at MessageStream.Writable.write (_stream_writable.js:318:11)
    at Socket.ondata (_stream_readable.js:695:22)
    at Socket.emit (events.js:310:20)
    at addChunk (_stream_readable.js:286:12)
    at readableAddChunk (_stream_readable.js:268:9)
    at Socket.Readable.push (_stream_readable.js:209:10)
    at TCP.onStreamRead (internal/stream_base_commons.js:186:23) {
  ok: 0,
  code: 18,
  codeName: 'AuthenticationFailed'
}
```

The `uri` would be `mongodb://root:password@localhost:27017/mydb?authSource=admin`

**Comment by hariprasanths** [comment]
AuthSource for the connection defaults to admin(no option to change it via the configuration for now).
It seems like you are trying to run this tool for windows, which isn't supported.
Tracking it here - #4

**Comment by bwgjoseph** [comment]
Sure, thanks for the clarification. Hope that configuration can be supported in the further. Thanks!

**Comment by bwgjoseph** [comment]
Hi @hariprasanths, I'm trying out this tool again after #4 and it seem like this issue is coming back. Would you revisit this? Thanks

**Comment by hariprasanths** [comment]
Can you also give me the log from the mongo daemon logs?

**Comment by bwgjoseph** [comment]
I think this is it

```json
{"t":{"$date":"2020-10-03T12:50:10.003+00:00"},"s":"I", "c":"NETWORK", "id":51800, "ctx":"conn113","msg":"client metadata","attr":{"remote":"172.29.0.1:33652","client":"conn113","doc":{"driver":{"name":"nodejs|Mongoose","version":"3.6.2"},"os":{"type":"Windows_NT","name":"win32","architecture":"x64","version":"10.0.18362"},"platform":"'Node.js v12.16.3, LE (unified)","version":"3.6.2|5.10.7"}}}

{"t":{"$date":"2020-10-03T12:50:10.012+00:00"},"s":"I", "c":"NETWORK", "id":22943, "ctx":"listener","msg":"connection accepted","attr":{"remote":"172.29.0.1:33656","sessionId":114,"connectionCount":6}}

{"t":{"$date":"2020-10-03T12:50:10.014+00:00"},"s":"I", "c":"NETWORK", "id":51800, "ctx":"conn114","msg":"client metadata","attr":{"remote":"172.29.0.1:33656","client":"conn114","doc":{"driver":{"name":"nodejs|Mongoose","version":"3.6.2"},"os":{"type":"Windows_NT","name":"win32","architecture":"x64","version":"10.0.18362"},"platform":"'Node.js v12.16.3, LE (unified)","version":"3.6.2|5.10.7"}}}

{"t":{"$date":"2020-10-03T12:50:10.014+00:00"},"s":"I", "c":"ACCESS", "id":20251, "ctx":"conn114","msg":"Supported SASL mechanisms requested for unknown user","attr":{"user":"root@feathers"}}

{"t":{"$date":"2020-10-03T12:50:10.014+00:00"},"s":"I", "c":"ACCESS", "id":20249, "ctx":"conn114","msg":"Authentication failed","attr":{"mechanism":"SCRAM-SHA-256","principalName":"root","authenticationDatabase":"feathers","client":"172.29.0.1:33656","result":"UserNotFound: Could not find user \"root\" for db \"feathers\""}}

{"t":{"$date":"2020-10-03T12:50:10.017+00:00"},"s":"I", "c":"ACCESS", "id":20249, "ctx":"conn114","msg":"Authentication failed","attr":{"mechanism":"SCRAM-SHA-1","principalName":"root","authenticationDatabase":"feathers","client":"172.29.0.1:33656","result":"UserNotFound: Could not find user \"root\" for db \"feathers\""}}

{"t":{"$date":"2020-10-03T12:50:10.021+00:00"},"s":"I", "c":"NETWORK", "id":22944, "ctx":"conn113","msg":"connection ended","attr":{"remote":"172.29.0.1:33652","connectionCount":5}}

{"t":{"$date":"2020-10-03T12:50:10.029+00:00"},"s":"I", "c":"NETWORK", "id":22944, "ctx":"conn114","msg":"connection ended","attr":{"remote":"172.29.0.1:33656","connectionCount":4}}
```

If I connect with `authSource=admin`, this is the log

```json
{"t":{"$date":"2020-10-03T12:51:59.618+00:00"},"s":"I", "c":"NETWORK", "id":22943, "ctx":"listener","msg":"connection accepted","attr":{"remote":"172.29.0.1:33660","sessionId":115,"connectionCount":5}}

{"t":{"$date":"2020-10-03T12:51:59.637+00:00"},"s":"I", "c":"NETWORK", "id":51800, "ctx":"conn115","msg":"client metadata","attr":{"remote":"172.29.0.1:33660","client":"conn115","doc":{"driver":{"name":"nodejs|Mongoose","version":"3.5.9"},"os":{"type":"Windows_NT","name":"win32","architecture":"x64","version":"10.0.18362"},"platform":"'Node.js v12.16.3, LE (unified)","version":"3.5.9|5.9.25"}}}

{"t":{"$date":"2020-10-03T12:51:59.669+00:00"},"s":"I", "c":"ACCESS", "id":20250, "ctx":"conn115","msg":"Successful authentication","attr":{"mechanism":"SCRAM-SHA-256","principalName":"root","authenticationDatabase":"admin","client":"172.29.0.1:33660"}}

{"t":{"$date":"2020-10-03T12:51:59.679+00:00"},"s":"I", "c":"NETWORK", "id":22943, "ctx":"listener","msg":"connection accepted","attr":{"remote":"172.29.0.1:33664","sessionId":116,"connectionCount":6}}

{"t":{"$date":"2020-10-03T12:51:59.679+00:00"},"s":"I", "c":"NETWORK", "id":51800, "ctx":"conn116","msg":"client metadata","attr":{"remote":"172.29.0.1:33664","client":"conn116","doc":{"driver":{"name":"nodejs|Mongoose","version":"3.5.9"},"os":{"type":"Windows_NT","name":"win32","architecture":"x64","version":"10.0.18362"},"platform":"'Node.js v12.16.3, LE (unified)","version":"3.5.9|5.9.25"}}}

{"t":{"$date":"2020-10-03T12:51:59.684+00:00"},"s":"I", "c":"ACCESS", "id":20250, "ctx":"conn116","msg":"Successful authentication","attr":{"mechanism":"SCRAM-SHA-256","principalName":"root","authenticationDatabase":"admin","client":"172.29.0.1:33664"}}
```

Let me know if you need more info

**Comment by bwgjoseph** [comment]
Yeah, I made changes to your source, and verified that I can connect now if I pass in the `authSource` params.

Before:

![image](https://user-images.githubusercontent.com/37484538/94992259-4c2d6a00-05bb-11eb-9f12-bc89509e04f6.png)

After:

```js
let connect = async (dbConfig) => {
    let connectionUri = mongoUriBuilder({
        username: dbConfig.dbAuthUser,
        password: dbConfig.dbAuthPwd,
        host: dbConfig.dbHost,
        port: dbConfig.dbPort,
        database: `${dbConfig.dbName}?authSource=admin`, // modified this
    });

    let connRes = await mongoose.connect(connectionUri, {useNewUrlParser: true, useUnifiedTopology: true});
    let disConnRes = await mongoose.connection.close();
    return connRes;
};
```

![image](https://user-images.githubusercontent.com/37484538/94992248-3f107b00-05bb-11eb-932a-6c94b453dbe3.png)

**Comment by hariprasanths** [comment]
Cool, I'll make the authSource to be configurable by the user (during the db configuration)

**Comment by bwgjoseph** [comment]
Yeah, thanks!

I think when you pass in the params for dump and restore would also required if configured. Also note that (if I'm not wrong), `authSource` could be other than `admin`

Although, I'm not sure what other issue there is but I tested to run the dump and it stuck in this for hours

![image](https://user-images.githubusercontent.com/37484538/94995293-ff07c300-05cf-11eb-8dd6-09d310e97d14.png)

Hope it helps!

**Comment by hariprasanths** [comment]
Yeah, that's right. AuthSource needs to be added to the dump and restore commands as well. Regarding your configuration, I can see that the db backup is scheduled at 00:00 everyday, you can change the scheduled time by running `--config db` again

**Comment by bwgjoseph** [comment]
Is there a way to run the scheduled task immediately?

**Comment by hariprasanths** [comment]
Yeah, you can set the time for the question (Enter the time to run the backups everyday?) In the db configuration to a closest time to check.

**Comment by bwgjoseph** [comment]
Yeah, I mean after I have scheduled it. Sometimes I would like to run it `on demand` rather than `on schedule`, it's easier that way than to keep changing my schedule. I guess that's for another topic/feature request (if there is no such thing now).

**Comment by hariprasanths** [comment]
That's not possible right now, you mean to say that you want to be able to do a instant backup for the day that has nothing to do with the schedule?

**Comment by bwgjoseph** [comment]
Yup, that's right. To have the ability to run the scheduled task on demand rather than waiting until the scheduled time for it to run

I would imagine something like `synchly --run`, maybe with ability to pass in `job name` as well. I could open another issue, so not to mix the concern if you like

**Comment by hariprasanths** [comment]
Yeah, that would be great.

**Comment by bwgjoseph** [comment]
HI @hariprasanths,

I see that a couple of issues that I raised has been supported but there isn't any new release, so I am unable to test those features. Any plans to release new version?

Thanks

**Comment by hariprasanths** [comment]
Yeah, we were waiting for another feature to be merged before the next release, but that could take a while.
We can do an intermediate release by today.
Will let you know when it is done.

**Comment by bwgjoseph** [comment]
I'm alright to wait for few weeks if it disrupt your release process or plan. No worries. 

Thanks for your support.
